kyuubi-server/src/main/scala/org/apache/kyuubi/engine/flink/FlinkProcessBuilder.scala

```diff
@@ -37,43 +41,75 @@ class FlinkProcessBuilder(
-        "For more detail information on installing and configuring Flink, please visit " +
-        "https://kyuubi.apache.org/docs/latest/deployment/settings.html#environments")
-    }
-  }
-
   override protected def module: String = "kyuubi-flink-sql-engine"
 
   override protected def mainClass: String = "org.apache.kyuubi.engine.flink.FlinkSQLEngine"
 
   override protected def childProcEnv: Map[String, String] = conf.getEnvs +
     ("FLINK_HOME" -> FLINK_HOME) +
     ("FLINK_CONF_DIR" -> s"$FLINK_HOME/conf") +
-    ("FLINK_SQL_ENGINE_JAR" -> mainResource.get) +
-    ("FLINK_SQL_ENGINE_DYNAMIC_ARGS" ->
-      conf.getAll.filter { case (k, _) =>
-        k.startsWith("kyuubi.") || k.startsWith("flink.") ||
-          k.startsWith("hadoop.") || k.startsWith("yarn.")
-      }.map { case (k, v) => s"-D$k=$v" }.mkString(" "))
-
-  override protected def commands: Array[String] = Array(executable)
+    ("_FLINK_HOME_DETERMINED" -> s"1")
+
+  override protected def commands: Array[String] = {
+    val buffer = new ArrayBuffer[String]()
+    buffer += s"bash"
+    buffer += s"-c"
+    val commandStr = new StringBuilder()
+
+    commandStr.append(s"source $FLINK_HOME${File.separator}bin" +
+      s"${File.separator}config.sh && $executable")
+
+    // TODO: How shall we deal with proxyUser,
+    // user.name
+    // kyuubi.session.user
+    // or just leave it, because we can handle it at operation layer
+    commandStr.append(s" -D$KYUUBI_SESSION_USER_KEY=$proxyUser ")
+
+    // TODO: add Kyuubi.engineEnv.FLINK_ENGINE_MEMORY or kyuubi.engine.flink.memory to configure
+    // -Xmx5g
+    // java options
+    val confStr = conf.getAll.filter { case (k, _) =>
+      k.startsWith("kyuubi.") || k.startsWith("flink.") ||
+        k.startsWith("hadoop.") || k.startsWith("yarn.")
+    }.map { case (k, v) => s"-D$k=$v" }.mkString(" ")
+    commandStr.append(confStr)
+
+    commandStr.append(" -cp ")
+    val classpathEntries = new LinkedHashSet[String]
+    // flink engine runtime jar
+    mainResource.foreach(classpathEntries.add)
+    // flink sql client jar
+    val flinkSqlClientPath = Paths.get(FLINK_HOME)
+      .resolve("opt")
+      .toFile
+      .listFiles(new FilenameFilter {
+        override def accept(dir: File, name: String): Boolean = {
+          name.toLowerCase.startsWith("flink-sql-client")
+        }
+      }).head.getAbsolutePath
+    classpathEntries.add(flinkSqlClientPath)
+
+    // jars from flink lib
+    classpathEntries.add(s"$FLINK_HOME${File.separator}lib${File.separator}*")
+
+    // classpath contains flink configurations, default to flink.home/conf
+    classpathEntries.add(env.getOrElse("FLINK_CONF_DIR", s"$FLINK_HOME${File.separator}conf"))
+    // classpath contains hadoop configurations
+    env.get("HADOOP_CONF_DIR").foreach(classpathEntries.add)
+    env.get("YARN_CONF_DIR").foreach(classpathEntries.add)
+    env.get("HBASE_CONF_DIR").foreach(classpathEntries.add)
+    val hadoopClasspath = env.get("HADOOP_CLASSPATH")
```

Comment:
how about FLINK_HADOOP_CLASSPATH?

Reply:
> how about FLINK_HADOOP_CLASSPATH?

Do you mean the same as ENGINE_HIVE_EXTRA_CLASSPATH, ENGINE_TRINO_EXTRA_CLASSPATH？

Reply:
yes

Reply:
> yes

I think another pr is needed to handle ENGINE_FLINK_MEMORY, ENGINE_FLINK_JAVA_OPTIONS, ENGINE_FLINK_EXTRA_CLASSPATH